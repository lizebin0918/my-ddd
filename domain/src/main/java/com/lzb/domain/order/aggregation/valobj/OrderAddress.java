package com.lzb.domain.order.aggregation.valobj;

import java.io.Serializable;

import com.fasterxml.jackson.annotation.JsonCreator;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.jackson.Jacksonized;


/**
 * 订单地址实体<br/>
 * Created on : 2023-08-31 12:52
 * @author lizebin
 */
@Getter
@Setter(AccessLevel.PACKAGE)
@Jacksonized
@Builder(toBuilder = true)
public class OrderAddress implements Serializable {

    private FullName fullName;

    private FullAddressLine fullAddressLine;

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    /**
     * 国家
     */
    private String country;

    @JsonCreator
    public OrderAddress(
            FullName fullName,
            FullAddressLine fullAddressLine,
            String email,
            String phoneNumber,
            String country) {
        this.fullName = fullName;
        this.fullAddressLine = fullAddressLine;
        this.email = email;
        this.phoneNumber = phoneNumber;
        this.country = country;
    }

}
