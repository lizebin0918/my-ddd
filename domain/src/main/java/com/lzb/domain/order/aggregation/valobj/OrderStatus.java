package com.lzb.domain.order.aggregation.valobj;

import com.lzb.component.utils.enums.EnumValue;
import lombok.AllArgsConstructor;
import lombok.Getter;

/**
 * 订单状态<br/>
 * Created on : 2023-09-04 10:00
 * @author lizebin
 */
@Getter
@AllArgsConstructor
public enum OrderStatus implements EnumValue<Integer> {

    WAIT_PAY(0, "待支付"), PAID(1, "已支付"), CANCELED(2, "已取消");

    private final Integer value;
    private final String desc;

    public OrderStatus toCancel() {
        if (isCancel()) {
            throw new IllegalStateException("订单（明细）已取消");
        }
        return CANCELED;
    }

    public boolean isCancel() {
        return this == CANCELED;
    }

}