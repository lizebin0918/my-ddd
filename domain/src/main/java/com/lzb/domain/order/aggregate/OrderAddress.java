package com.lzb.domain.order.aggregate;

import cn.hutool.core.lang.Assert;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.domain.aggregate.BaseEntity;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;


/**
 * 订单地址实体<br/>
 * Created on : 2023-08-31 12:52
 * @author lizebin
 */
@Getter
// 方便测试构造
@Setter(AccessLevel.PACKAGE)
public class OrderAddress extends BaseEntity<OrderAddress> {

    private FullName fullName;

    private FullAddressLine fullAddressLine;

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    /**
     * 国家
     */
    private String country;

    protected OrderAddress(long id) {
        super(id);
    }

    @JsonCreator
    public OrderAddress(Long id,
            @NonNull FullName fullName,
            @NonNull FullAddressLine fullAddressLine,
            @NonNull String email,
            @NonNull String phoneNumber,
            @NonNull String country) {
        super(id);
        this.fullName = fullName;
        this.fullAddressLine = fullAddressLine;
        this.email = email;
        this.phoneNumber = phoneNumber;
        this.country = country;
    }

    /**
     * 更新姓名
     * @param fullName
     */
    void updateFullName(FullName fullName) {
        Assert.notBlank(fullName.getFirstName(), "firstName is null");
        this.fullName = fullName;
    }
}
