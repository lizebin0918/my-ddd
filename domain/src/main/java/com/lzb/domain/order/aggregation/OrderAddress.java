package com.lzb.domain.order.aggregation;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.domain.order.aggregation.valobj.FullAddressLine;
import com.lzb.domain.order.aggregation.valobj.FullName;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;


/**
 * 订单地址实体<br/>
 * Created on : 2023-08-31 12:52
 * @author lizebin
 */
@Getter
@Setter(AccessLevel.PACKAGE)
@Builder(toBuilder = true)
public class OrderAddress {

    private FullName fullName;

    private FullAddressLine fullAddressLine;

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    /**
     * 国家
     */
    private String country;

    @JsonCreator
    public OrderAddress(
            @NonNull FullName fullName,
            @NonNull FullAddressLine fullAddressLine,
            @NonNull String email,
            @NonNull String phoneNumber,
            @NonNull String country) {
        this.fullName = fullName;
        this.fullAddressLine = fullAddressLine;
        this.email = email;
        this.phoneNumber = phoneNumber;
        this.country = country;
    }

}
