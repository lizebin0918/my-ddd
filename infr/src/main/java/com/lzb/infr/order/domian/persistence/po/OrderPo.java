package com.lzb.infr.order.domian.persistence.po;

import java.math.BigDecimal;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.annotation.Version;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.infr.common.BasePo;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.javers.core.metamodel.annotation.Id;

/**
 * <p>
 * 
 * </p>
 *
 * @author lizebin
 * @since 2023-08-29
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("\"order\"")
public class OrderPo extends BasePo {

    @Id
    @TableId(type = IdType.INPUT)
    private Long orderId;

    private OrderStatus orderStatus;

    private String currency;

    /**
     * 汇率
     */
    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    @Version
    private int version;

    ///////////////////////////////////////////////////////////////////////////
    // 地址相关
    ///////////////////////////////////////////////////////////////////////////

    /**
     * 邮箱
     */
    private String email;

    /**
     * 电话
     */
    private String phoneNumber;

    private String firstName;

    private String lastName;

    /**
     * 地址1
     */
    private String addressLine1;

    /**
     * 地址2
     */
    private String addressLine2;

    /**
     * 国家
     */
    private String country;
}