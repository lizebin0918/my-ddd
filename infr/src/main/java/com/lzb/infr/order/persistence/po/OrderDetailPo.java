package com.lzb.infr.order.persistence.po;

import java.math.BigDecimal;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.infr.common.BasePo;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.javers.core.metamodel.annotation.Id;

@Data
@EqualsAndHashCode(callSuper = false)
@TableName("order_detail")
public class OrderDetailPo extends BasePo {

    @Id
    @TableId(type = IdType.AUTO)
    private Long id;

    private Long orderId;

    private Integer skuId;

    private OrderStatus orderStatus;

    private BigDecimal price;

    private Boolean locked;


}
