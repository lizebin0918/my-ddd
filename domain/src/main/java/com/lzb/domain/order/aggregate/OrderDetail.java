package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.domain.aggregate.BaseEntity;
import com.lzb.domain.order.enums.OrderStatus;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;

/**
 * 订单明细<br/>
 * Created on : 2023-08-31 12:51
 * @author lizebin
 */
@Slf4j
@Getter
// 方便测试构造
@Setter(AccessLevel.PACKAGE)
public class OrderDetail extends BaseEntity<OrderDetail> {

    private int skuId;

    private OrderStatus orderStatus;

    private BigDecimal price;

    /**
     * null:未知
     * true:锁定库存
     * false:缺货
     */
    private Boolean locked;

    protected OrderDetail(long id) {
        super(id);
    }

    /**
     * 直接提供给仓储层使用，无需校验业务逻辑
     * @param id
     * @param orderId
     * @param skuId
     * @param orderStatus
     * @param price
     */
    @JsonCreator
    public OrderDetail(long id, int skuId, @NonNull OrderStatus orderStatus, @NonNull BigDecimal price, Boolean locked) {
        super(id);
        this.skuId = skuId;
        this.orderStatus = orderStatus;
        this.price = price;
        this.locked = locked;
    }

    /**
     * 订单明细取消
     */
    void cancel() {
        orderStatus = orderStatus.toCancel();
    }

    /**
     * 更新库存锁定状态
     * @param locked
     */
    void updateLocked(Boolean locked) {
        this.locked = locked;
    }
}
