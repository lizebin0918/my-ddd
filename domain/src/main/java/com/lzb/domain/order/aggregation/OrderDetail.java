package com.lzb.domain.order.aggregation;

import java.math.BigDecimal;

import com.lzb.component.domain.aggregate.BaseEntity;
import com.lzb.domain.order.aggregation.valobj.OrderStatus;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;

/**
 * 订单明细<br/>
 * Created on : 2023-08-31 12:51
 * @author lizebin
 */
@Slf4j
@Getter
@Jacksonized
@SuperBuilder
@Setter(AccessLevel.PACKAGE)
public class OrderDetail extends BaseEntity<OrderDetail> {

    private int skuId;

    private OrderStatus orderStatus;

    private BigDecimal price;

    /**
     * null:未知
     * true:锁定库存
     * false:缺货
     */
    private Boolean locked;

    /**
     * 订单明细取消
     */
    void cancel() {
        orderStatus = orderStatus.toCancel();
    }

    /**
     * 更新库存锁定状态
     * @param locked
     */
    void updateLocked(Boolean locked) {
        this.locked = locked;
    }
}
