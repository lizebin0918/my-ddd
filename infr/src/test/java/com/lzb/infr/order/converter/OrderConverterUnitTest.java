package com.lzb.infr.order.converter;

import java.math.BigDecimal;

import com.lzb.BaseUnitTest;
import com.lzb.domain.order.aggregation.OrderAddress;
import com.lzb.domain.order.aggregation.OrderDetail;
import com.lzb.domain.order.aggregation.valobj.OrderStatus;
import com.lzb.infr.order.persistence.po.OrderDetailPo;
import com.lzb.infr.order.persistence.po.OrderPo;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class OrderConverterUnitTest extends BaseUnitTest {

    @org.junit.jupiter.api.Test
    void toOrder() {
    }

    @org.junit.jupiter.api.Test
    void toOrderDetails() {
    }

    @Test
    @DisplayName("测试字段映射")
    void toOrderDetail() {

        OrderDetailPo orderDetailPo = new OrderDetailPo();
        orderDetailPo.setId(0L);
        orderDetailPo.setOrderId(0L);
        orderDetailPo.setSkuId(0);
        orderDetailPo.setOrderStatus(OrderStatus.WAIT_PAY);
        orderDetailPo.setPrice(BigDecimal.TEN);

        OrderDetail orderDetail = OrderConverter.toOrderDetail(orderDetailPo);
        assertJSON(orderDetail);
    }

    @Test
    @DisplayName("测试订单地址字段映射")
    void toOrderAddress() {

        OrderPo orderPo = new OrderPo();
        orderPo.setOrderId(888L);
        orderPo.setEmail("thisIsEmail@qq.com");
        orderPo.setPhoneNumber("123456");
        orderPo.setFirstName("firstName");
        orderPo.setLastName("lastName");
        orderPo.setAddressLine1("addressLine1");
        orderPo.setAddressLine2("addreaaLine2");
        orderPo.setCountry("country");

        OrderAddress orderAddress = OrderConverter.toOrderAddress(orderPo);
        assertJSON(orderAddress);

    }
}