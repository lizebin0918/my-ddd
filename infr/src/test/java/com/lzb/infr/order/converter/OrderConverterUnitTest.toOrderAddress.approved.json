{
  "country": "country",
  "email": "thisIsEmail@qq.com",
  "fullAddressLine": {
    "addressLine1": "addressLine1",
    "addressLine2": "addreaaLine2"
  },
  "fullName": {
    "firstName": "firstName",
    "lastName": "lastName"
  },
  "phoneNumber": "123456"
}